test/property_test.c

```diff
@@ -0,0 +1,369 @@
+        && TEST_int_eq(ossl_property_value("yes", 1), j)
+        && TEST_int_eq(ossl_property_value("no", 1), i)
+        && TEST_int_ne(i = ossl_property_value("green", 1), 0)
+        && TEST_int_eq(k = ossl_property_value("fnord", 1), i + 1)
+        && TEST_int_eq(ossl_property_value("fnord", 1), k)
+        && TEST_str_eq(ossl_property_value_from_idx(i), "green")
+        && TEST_str_eq(ossl_property_value_from_idx(k), "fnord")
+        /* Check name and values are distinct */
+        && TEST_int_eq(ossl_property_value("cold", 0), 0)
+        && TEST_int_ne(ossl_property_name("fnord", 0),
+                       ossl_property_value("fnord", 0)))
+        res = 1;
+    ossl_property_string_cleanup();
+    return res;
+}
+
+static const struct {
+    const char *defn;
+    const char *query;
+    int e;
+} parser_tests[] = {
+    { "", "sky=blue", 0 },
+    { "", "sky!=blue", 1 },
+    { "green", "", 1 },
+    { "cold=yes", "cold=yes", 1 },
+    { "cold=yes", "cold", 1 },
+    { "cold=yes", "cold!=no", 1 },
+    { "green", "green=yes", 1 },
+    { "green", "green=no", 0 },
+    { "green", "green!=yes", 0 },
+    { "cold=no", "cold", 0 },
+    { "cold=no", "cold=no", 1 },
+    { "green", "cold", 0 },
+    { "green", "cold=no", 1 },
+    { "green", "cold!=yes", 1 },
+    { "green=blue", "green=yellow", 0 },
+    { "green=blue", "green!=yellow", 1 },
+    { "today=monday, tomorrow=3", "today!=2", 1 },
+    { "today=monday, tomorrow=3", "today!='monday'", 0 },
+    { "today=monday, tomorrow=3", "tomorrow=3", 1 },
+    { "n=0x3", "n=3", 1 },
+    { "n=0x3", "n=-3", 0 },
+    { "n=0x33", "n=51", 1 },
+    { "n=033", "n=27", 1 },
+    { "n=0", "n=00", 1 },
+    { "n=0x0", "n=0", 1},
+};
+
+static int test_property_parse(int n)
+{
+    PROPERTY_LIST *p = NULL, *q = NULL;
+    int r = 0;
+
+    if (TEST_true(ossl_property_string_init())
+        && TEST_true(ossl_property_parse_init())
+        && add_property_names("sky", "green", "cold", "today", "tomorrow", "n",
+                              NULL)
+        && TEST_ptr(p = ossl_parse_property(parser_tests[n].defn))
+        && TEST_ptr(q = ossl_parse_query(parser_tests[n].query))
+        && TEST_int_eq(ossl_property_compare(q, p), parser_tests[n].e))
+        r = 1;
+    ossl_property_free(p);
+    ossl_property_free(q);
+    ossl_property_string_cleanup();
+    return r;
+}
+
+static const struct {
+    const char *a1;
+    const char *a2;
+    const char *d;
+} merge_tests[] = {
+    { "colour=blue", "", "colour=blue" },
+    { "", "colour=blue", "colour=blue" },
+    { "colour=blue", "colour=red", "colour=blue" },
+    { "sky=blue, colour=green", "clouds=pink, sky=red",
+        "sky=blue, colour=green, clouds=pink" },
+    { "sky=blue", "pot=gold", "pot=gold, sky=blue" },
+    { "day", "night", "day=yes, night=yes" },
+    { "night", "day", "day=yes, night=yes" },
+    { "", "", "" },
+    { "-day", "day=yes", "day=no" },
+    { "-night, day", "day, night", "day=yes, night=no" },
```

Comment:
I find these last two lines unexpected. The design doc has this example:
Global setting: fips=no
Passed setting: -fips
Resulting query: fips is not specified

Therefore I would have expected the above two lines to be:
````
    { "-day", "day=yes", "" },
    { "-night, day", "day, night", "day=yes" },
````

Reply:
I agree with @mattcaswell here

Reply:
Is it the ordering of the fields that the problem?  They are ordered _local_, _global_ and _result_.  If I switch them for _global_, _local_ then _result_, it would be okay?  Did the `"colour=blue", "colour=red", "colour=blue"` earlier confuse too?

Or is it the =no tests (meaning test for undefined or "no")?

Reply:
Yep the ordering is confusing, Switch them and maybe change the struct names to make it more obvious.

Reply:
Ordering switched and better names allocated.

Reply:
For me, it was that the combo of "-foo" and "foo=yes" would give "foo=no". As far as I understood, the result should be nothing, i.e. "foo" unspecified.

Reply:
We can't query for "unspecified", just "yes" or "no".  "foo=no" is true if unspecified however (both being false).

I think this is the corrent behaviour, since no provider will set everything it doesn't do to "no" but it will set what it does do to "yes".

Reply:
> { "-day", "day=yes", "" },

Noting that the empty query here would match everything regardless of what is set and what isn't.  This isn't what needs to be validated as working at this point.

Reply:
Uhmmm, so here's the examples in the design document, under "Multileve Queries":

-   global: fips=yes
    passed: -fips
    resulting: fips is not specified
-   global: fips=no
    passed: -fips
    resulting: fips is not specified

Could you tell us how that corresponds to what you just said?

Reply:
Hmmm. To me "fips=no" means do not select an algorithm defined as "fips=yes". Whereas "" means I don't care whether you select an algorithm with "fips=yes" defined or not.

Reply:
...and so of course global "fips=yes" and passed "-fips" means I don't care whether you select fips or not - which is not the same as "fips=no".

Reply:
I think the disconnect here is that these are _test_ cases, not _examples_ of usage.

How about I add another case "fips=yes" globally, "-fips" locally and "fips=yes" as the definition?  Or "fips=yes" then "-fips" and a definition of "sesquioxidizing=1960"?

Reply:
I've added some additional test cases for the '-' prefix.

Reply:
This helps immensely! It suddenly all makes sense to me. Actually I think the source of the confusion (in part) is the function name `ossl_property_compare`. One intuitively expects that function to take two property lists and return true if they are equal. Hence merging local "-day" with global "day=yes" and saying that is equal to "day=no" is not what I expected. Actually it doesn't "compare" lists exactly at all. It tells you whether a query matches against a property definition. Perhaps a different name for the function?